stable/fluent-bit/templates/config.yaml

```diff
@@ -31,19 +31,20 @@ data:
         Tag              {{ .Values.filter.kubeTag }}.*
         Refresh_Interval 5
         Mem_Buf_Limit    {{ .Values.input.tail.memBufLimit }}
-        Skip_Long_Lines  On
+        Skip_Long_Lines  {{ .Values.input.tail.skipLongLines }}
 {{- if .Values.trackOffsets }}
         DB               /tail-db/tail-containers-state.db
         DB.Sync          Normal
 {{- end }}
-{{ .Values.extraInputs | indent 8 }}
+{{ .Values.extraInputs | indent 4 }}
```

Comment:
Why is this here? This will break things.

Reply:
Infact, there was a recent patch, which I was confusing this one with, that merged fairly recently:
https://github.com/helm/charts/commit/9ab3cbbc28bf0416cf5400fb9ea4679763807506

Reply:
[9ab3cbb](https://github.com/helm/charts/commit/9ab3cbbc28bf0416cf5400fb9ea4679763807506) is incorrect because it is trying to add additional parameters/options to the existing plugin. my understanding is `extraInputs` is used for adding addition `[INPUT]` section and so on with `extraOutputs` and `extraFilters`. Maybe a separate variable is needed to implement [9ab3cbb](https://github.com/helm/charts/commit/9ab3cbbc28bf0416cf5400fb9ea4679763807506).

Reply:
@kfox1111 looking at the readme, it is clear that those variables are there for adding additional sections, not adding config values to existing sections:

```

extraInputs | Add extra Input sections to config | ``
-- | -- | --
extraFilters | Add extra Filter sections to config | ``
extraOutputs | Add extra Output sections to config | ``
```

Reply:
yeah, I think its correct in that the idea was to let the user add more options to the existing input/filter/outputs.

I don't think toml lets you have multiple sections with the same name? so it won't ever work I think.

Reply:
Yeah, I think that readme could probably be worded better. its sections as in lines of text, not sections as in a new block of toml. Any idea how to more concisely wordsmith that?

Reply:
Oh yes, it does,  please see the sections doc here, https://docs.fluentbit.io/manual/configuration/schema#sections, 

> Multiples sections can exists on the same file.

And a example of this filter sections: https://docs.fluentbit.io/manual/filter/nest#configuration-file-3

thanks!

Reply:
I think the readme is correct, if you take the `extraOutputs` it is added at the end of all the outputs sections, so this would mean that only `http` output can have additional parameters and the other output sections file, forward, es and splunk cannot. 

also the extras{Inputs,Ouputs,Filters} were added as part of https://github.com/helm/charts/pull/8600, which clearly states the intent was to add additional parameters. It was misused and then an attempt to patch in 9ab3cbb.

Reply:
oh gosh... that feels really weird... only sax style parsers would be able to handle it. Dom parsers will be inconsistent or break. this parser chokes on that example:
http://toml-online-parser.ovonick.com/

Ok, well, if its a design feature of fluent-bit, then it is what it is. :/

Do the sections all need to be in order? If not, I'd recommend leaving these options alone, and maybe adding an .extraSections option towards the bottom?

Reply:
Also the author of that PR admits it was not intended for that purpose and the fix was not correct: https://github.com/helm/charts/pull/8825#issuecomment-435867527, but it was never reverted. and another PR has been opened to fix the same issue: https://github.com/helm/charts/pull/9008.

Reply:
I am not sure if the sections have to be in order, docs don't mention it, https://docs.fluentbit.io/manual/configuration/schema#sections. So I am not sure.

Reply:
also this comment addresses how the same variable can be used to have a section and a parameter for the section above if needed: https://github.com/helm/charts/pull/9008#issuecomment-439481904.

Reply:
hmm.. yeah. that pr does seem to have intended it to be for extra sections... I never reviewed it though. I think I would have likely blocked it in that form, asking the same questions I am now...

shoehorning yaml into toml like this is really kind of gross. I'm kind of of the opinion, make the 20% of the config options that 80% of the users need super easy to set and for the rest, just let the user upload their own fully customized configmap via existingConfigMap. Otherwise we're going to keep adding mapping after mapping of stuff from yaml to toml. Have we reached that point here?

I get making Skip_Long_Lines configurable since it can't be overridden otherwise.

Multiple INPUTS/FILTERS/OUTPUT sections feels like its over the line though?

Reply:
Well, the thing is you have to take into consideration the application, in this case fluent-bit, I really only want to (in most cases) change the INPUT/FILTERS/OUTPUT sections. And change the logs to a format I wanted and send it to a backend. 

Now if I have to overwrite a configmap, I to write the whole config file for fluent-bit and then apply it. 

Right now with this simple config, it works.

```
fluent-bit:
  backend:
    forward:
      host: logging
      port: 24224
  extraFilters: |-

    [FILTER]
        Name nest
        Match kube.*
        Operation lift
        Nested_under kubernetes
        Add_prefix kubernetes_
```

Reply:
https://github.com/helm/charts/pull/3573

Also the implementation of overriding configmap is done weirdly, so I have to have a separate process create the configmap and then have that configmap name provided in the chart values file. And that configmap is out of scope for helm to manage, update, etc. I think the configmap override should have been done at the `data` key at https://github.com/helm/charts/blob/master/stable/fluent-bit/templates/config.yaml#L11, so I could have provided the configmap in the values file. I do not want to manage kubernetes objection creation separately, helm and charts are made to do exactly that.

Reply:
The issue I have is that to add extra options to say the existing filter, then you have to know how much whitespace is involved which should be an implementation detail that should never be exposed.

I think your example could be done like:
```yaml
fluent-bit:
  backend:
    forward:
      host: logging
      port: 24224
  extraSections: |-

    [FILTER]
        Name nest
        Match kube.*
        Operation lift
        Nested_under kubernetes
        Add_prefix kubernetes_
```

if this were added at the end of the configmap:
```
{{ .Values.extraSections | indent 4 }}
```
It would allow adding to the existing values without needing to know the implementation detail of how far indented the config needs to be, I'd be ok with that. Could you test that?

Reply:
sure, testing now.

Reply:
Just going by this example, filters might have to be in sequence: https://docs.fluentbit.io/manual/filter/nest#configuration-file-3

Reply:
It might work in my test, because I am just adding filters, but it might break for others, who need the correct sequence.

Reply:
@edsiper can you please weigh in on if the config sections will always work out of order?

Reply:
In general, there's a reason why configmaps are text blobs in k8s and not yaml or something. The k8s developers realized a common problem in config management is that config managers try and make the config standardized into some format (say yaml) and then map it into the application's config. Since there is a million ways to write out configs, there is a million ways to convert the input to the output. This leads to a snowballing of config management code, unit tests, etc. Its a painful treadmill. k8s put this out of scope and said, if you want to pass config to sendmail in m4 macro's, then stick it in a configmap and we'll ship it without touching it. It will just work. Helm suffers from the config mapping issue quite a bit. The more config options you want to have in values, the more conditional/gotl template crazyness goes on until the template is unreadable. By allowing the user to upload a configmap directly, that gets us off the treadmill of supporting every possible option in every possible combination. If you fit in the 80% common case, values is easy and there. If you don't you upload your config and it supports every possible option you might need, without the chart becoming unreadable/unmaintainable. This feels like a good tradeoff of developer and user time.

Reply:
As it stands, I can't even do that, the override configmap options is for the configmap to be created separately, outside of the helm chart entirely.

Reply:
Yup. Do:
```
kubectl create configmap fluent-bit --from-file=fluent-bit.conf
```

Reply:
adding the sections at the end worked for me, but not certain if it will work in all scenarios. doing a test with nesting and putting the filters out of order to see if that works as intended.

Reply:
Yeah, well I understand how to make a configmap from a file. That is not at all the issue, that would break the automation we have, which expects logging to be installed via charts entirely.

Reply:
Another option is to throw the configmap into your own chart. and possibly add the fluent-bit one as a subchart. At some point, you gotta maintain your own things. But this allows you to share as much as possible. If you have to do absolutely everything through helm its likely you will need an integration chart like that at some point.

I'm really worried about the current templates getting unmaintainable. We've broken them lots of times already adding options. The more we add, the more likely it is to be broken again.

Reply:
FYI, i was talking with donbowman and he mentioned actually, all of the following are potentially valid and different configs:
 * [input] [input] [filter] [filter] [output] [output]
 * [input] [filter] [output] [input] [filter] [output]
and even orderings of the filters can be different with different results...
 * [input] [filter a] [filter b] [output]
 * [input] [filter b] [filter a] [output]

So, extraInputs/extraOutputs doesn't really make too much sense at 4 indentions.  extraFilters kind of does, but you may want a preExtraFilters too then, and you still might want an extraSections.

All this to say, its super complicated making the values flexible enough to cover all cases. Instead, lets just cover the [input][filter][output] use case, and use your own configmap if you must differ.

Reply:
ok cool. I used the approach you suggested and went with installing the configmap using a parent/bridge chart (which we already had) and fluent-bit as a dependency. worked out nicely! thanks @kfox1111!

Reply:
yes, they are all valid, filters once has to be careful of, because like you stated and with my testing, the sequence matters to get the results one wants.

Reply:
Ok, cool. glad its working for you. :)

Reply:
The key is called `extraInputs`. I wanted to add _extra inputs_ to my configmap, so I tried this key and it didn't work!

Also, `indent 4` is far more flexible than `indent 8` because I can self indent 4 to achieve indent 8 if I want to add entries to an existing INPUT section, but I cant negatively indent to add an entire new INPUT section.